pkg/controller/installation/bpf.go

```diff
@@ -0,0 +1,170 @@
+
+	var patchFelixConfig bool
+
+	// Check the install dataplane mode is either Iptables or BPF.
+	installBpfEnabled := common.BpfDataplaneEnabled(&install.Spec)
+
+	// Check edge case where User has externally patched FelixConfig bpfEnabled which causes conflict to prevent Operator from upgrading dataplane.
+	if fc.Spec.BPFEnabled != nil {
+
+		fcBPFEnabled := *fc.Spec.BPFEnabled
+		if installBpfEnabled != fcBPFEnabled {
+
+			// Ensure Felix Config annotations are either empty or equal previous FC bpfEnabled value.
+			if fc.Annotations[render.BpfOperatorAnnotation] == strconv.FormatBool(installBpfEnabled) {
+				text := fmt.Sprintf("An error occurred while attempting patch Felix Config bpfEnabled: '%s' as Felix Config has been modified externally to '%s'",
+					strconv.FormatBool(installBpfEnabled),
+					strconv.FormatBool(fcBPFEnabled))
+				err := errors.New(text)
+				return err
+			}
+		}
+	}
+
+	if !installBpfEnabled {
+		// IP Tables dataplane:
+		// Only patch Felix Config once to prevent log spamming.
+		if fc.Spec.BPFEnabled == nil || *fc.Spec.BPFEnabled {
+			patchFelixConfig = true
+		}
+	} else {
+		// BPF dataplane:
+		// Check daemonset rollout complete before patching.
+		if fc.Spec.BPFEnabled == nil || !(*fc.Spec.BPFEnabled) {
+			patchFelixConfig = checkDaemonsetRolloutComplete(ds)
+		}
+	}
+
+	// Attempt to patch Felix Config now.
+	return patchFelixConfigurationImpl(r, ctx, install, fc, reqLogger, patchFelixConfig)
+}
+
+func processDaemonsetEnvVar(r *ReconcileInstallation, ctx context.Context, ds *appsv1.DaemonSet, fc *crdv1.FelixConfiguration, reqLogger logr.Logger) (bool, error) {
+
+	dsBpfEnabledEnvVar, err := convert.GetEnv(ctx, r.client, ds.Spec.Template.Spec, convert.ComponentCalicoNode, common.NodeDaemonSetName, "FELIX_BPFENABLED")
+	if err != nil {
+		reqLogger.Error(err, "An error occurred when querying Calico-Node environment variable FELIX_BPFENABLED")
+		return false, err
+	}
+
+	dsBpfEnabledStatus := false
+	if dsBpfEnabledEnvVar != nil {
+		dsBpfEnabledStatus, err = strconv.ParseBool(*dsBpfEnabledEnvVar)
+		if err != nil {
+			reqLogger.Error(err, "An error occurred when converting Calico-Node environment variable FELIX_BPFENABLED")
+			return false, err
+		}
+	}
+
+	return dsBpfEnabledStatus && fc.Spec.BPFEnabled == nil, nil
+}
+
+func checkDaemonsetRolloutComplete(ds *appsv1.DaemonSet) bool {
+
+	if ds.Spec.Template.Spec.Volumes == nil {
+		return false
+	}
+
+	for _, volume := range ds.Spec.Template.Spec.Volumes {
+		if volume.Name == common.BPFVolumeName {
+			return ds.Status.CurrentNumberScheduled == ds.Status.UpdatedNumberScheduled && ds.Status.CurrentNumberScheduled == ds.Status.NumberAvailable
+		}
+	}
+
+	return false
+}
+
+func patchFelixConfigurationImpl(r *ReconcileInstallation, ctx context.Context, install *operator.Installation, fc *crdv1.FelixConfiguration, reqLogger logr.Logger, patchFelixConfig bool) error {
```

Comment:
Why would a caller ever want to call this function with `patchFelixConfig` set to `false` - it effectively makes the function a no-op. The caller should just not call this function if it doesn't want to patch the felix configuration.

Reply:
Done - thanks